google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/SpannerPool.java

```diff
@@ -0,0 +1,407 @@
+          });
+    }
+    return builder.build().getService();
+  }
+
+  /**
+   * Remove the given {@link ConnectionImpl} from the list of connections that should be monitored
+   * by this pool.
+   *
+   * @param options The {@link ConnectionOptions} that were used to create the connection.
+   * @param connection The {@link ConnectionImpl} to remove from this pool..
+   */
+  void removeConnection(ConnectionOptions options, ConnectionImpl connection) {
+    Preconditions.checkNotNull(options);
+    Preconditions.checkNotNull(connection);
+    SpannerPoolKey key = SpannerPoolKey.of(options);
+    synchronized (this) {
+      if (spanners.containsKey(key) && connections.containsKey(key)) {
+        List<ConnectionImpl> registeredConnections = connections.get(key);
+        // Remove the connection from the pool.
+        if (registeredConnections == null || !registeredConnections.remove(connection)) {
+          logger.log(
+              Level.WARNING,
+              "There are no connections registered for ConnectionOptions " + options.toString());
+        } else {
+          // Check if this was the last connection for this spanner key.
+          if (registeredConnections.isEmpty()) {
+            // Register the moment the last connection for this Spanner key was removed, so we know
+            // which Spanner objects we could close.
+            lastConnectionClosedAt.put(key, System.currentTimeMillis());
+          }
+        }
+      } else {
+        logger.log(
+            Level.WARNING,
+            "There is no Spanner registered for ConnectionOptions " + options.toString());
+      }
+    }
+  }
+
+  /**
+   * Checks that there are no {@link Connection}s that have been created by this {@link SpannerPool}
+   * that are still open, and then closes all {@link Spanner} instances in the pool. If there is at
+   * least one unclosed {@link Connection} left in the pool, the method will throw a {@link
+   * SpannerException} and no {@link Spanner} instances will be closed.
+   */
+  void checkAndCloseSpanners() {
+    checkAndCloseSpanners(CheckAndCloseSpannersMode.ERROR);
+  }
+
+  @VisibleForTesting
+  void checkAndCloseSpanners(CheckAndCloseSpannersMode mode) {
+    List<SpannerPoolKey> keysStillInUse = new ArrayList<>();
+    synchronized (this) {
+      for (Entry<SpannerPoolKey, Spanner> entry : spanners.entrySet()) {
+        if (!lastConnectionClosedAt.containsKey(entry.getKey())) {
+          keysStillInUse.add(entry.getKey());
+        }
+      }
+      if (keysStillInUse.isEmpty() || mode == CheckAndCloseSpannersMode.WARN) {
+        if (!keysStillInUse.isEmpty()) {
+          logLeakedConnections(keysStillInUse);
+          logger.log(
+              Level.WARNING,
+              "There is/are "
+                  + keysStillInUse.size()
+                  + " connection(s) still open."
+                  + " Close all connections before stopping the application");
+        }
+        closeUnusedSpanners(Long.MIN_VALUE);
```

Comment:
to confirm my understanding, passing through Long.MIN_VALUE will force the spanner sessions to close?

Reply:
Correct. I've added a comment to the method invocation to make that clearer.